File: backend/src/AnimeShop/API/Maintenance/EquipmentInspectionSchedules.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
module AnimeShop.API.Maintenance.EquipmentInspectionSchedules where

import AnimeShop.Prelude

type EquipmentInspectionSchedulesAPI =
  "equipment_inspection_schedules" :>
  (
    Get '[JSON] [Entity EquipmentInspectionSchedule] :<|>
    ReqBody '[JSON] EquipmentInspectionSchedule :> Post '[JSON] (Entity EquipmentInspectionSchedule) :<|>
    ReqBody '[JSON] EquipmentInspectionScheduleId :> Delete '[JSON] () :<|>
    ReqBody '[JSON] (Entity EquipmentInspectionSchedule) :> Patch '[JSON] ()
  )

equipmentInspectionSchedulesServer :: ServerT EquipmentInspectionSchedulesAPI App
equipmentInspectionSchedulesServer = getEquipmentInspectionSchedules :<|> postEquipmentInspectionSchedules :<|> deleteEquipmentInspectionSchedule :<|> patchEquipmentInspectionSchedule

getEquipmentInspectionSchedules :: App [Entity EquipmentInspectionSchedule]
getEquipmentInspectionSchedules = runDb $ selectList [] []

postEquipmentInspectionSchedules :: EquipmentInspectionSchedule -> App (Entity EquipmentInspectionSchedule)
postEquipmentInspectionSchedules = runDb . insertEntity

deleteEquipmentInspectionSchedule :: EquipmentInspectionScheduleId -> App ()
deleteEquipmentInspectionSchedule = runDb . delete

patchEquipmentInspectionSchedule :: Entity EquipmentInspectionSchedule -> App ()
patchEquipmentInspectionSchedule entity = runDb $ replace (entityKey entity) (entityVal entity)

File: frontend/src/reducers/maintenance/equipmentInspectionSchedules.js
import { ADD_EQUIPMENT_INSPECTION_SCHEDULE, GET_EQUIPMENT_INSPECTION_SCHEDULES, DELETE_EQUIPMENT_INSPECTION_SCHEDULE } from '../../actions/types';

export function equipmentInspectionSchedules(state = [], action) {
  switch (action.type) {
    case ADD_EQUIPMENT_INSPECTION_SCHEDULE:
      return [ ...state, action.equipmentInspectionSchedule];
    case GET_EQUIPMENT_INSPECTION_SCHEDULES:
      return action.equipmentInspectionSchedules;
    case DELETE_EQUIPMENT_INSPECTION_SCHEDULE:
      return state.filter(equipmentInspectionSchedule => equipmentInspectionSchedule.id !== action.id);
    default:
      return state;
  }
}
